include/zephyr/fs/zms.h

```diff
@@ -0,0 +1,211 @@
+/*  ZMS: Zephyr Memory Storage
+ *
+ * Copyright (c) 2024 BayLibre SAS
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_INCLUDE_FS_ZMS_H_
+#define ZEPHYR_INCLUDE_FS_ZMS_H_
+
+#include <sys/types.h>
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/toolchain.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Zephyr Memory Storage (ZMS)
+ * @defgroup zms Zephyr Memory Storage (ZMS)
+ * @ingroup file_system_storage
+ * @{
+ * @}
+ */
+
+/**
+ * @brief Zephyr Memory Storage Data Structures
+ * @defgroup zms_data_structures Zephyr Memory Storage Data Structures
+ * @ingroup zms
+ * @{
+ */
+
+/**
+ * @brief Zephyr Memory Storage File system structure
+ */
+struct zms_fs {
+	/** File system offset in flash **/
+	off_t offset;
+	/** Allocation table entry write address.
+	 * Addresses are stored as uint64_t:
+	 * - high 4 bytes correspond to the sector
+	 * - low 4 bytes are the offset in the sector
+	 */
+	uint64_t ate_wra;
+	/** Data write address */
+	uint64_t data_wra;
+	/** Storage system is split into sectors, each sector size must be multiple of erase-blocks
+	 *  if the device has erase capabilities
+	 */
+	uint32_t sector_size;
+	/** Number of sectors in the file system */
+	uint32_t sector_count;
+	/** Current cycle counter of the active sector (pointed by ate_wra)*/
+	uint8_t sector_cycle;
+	/** Flag indicating if the file system is initialized */
+	bool ready;
+	/** Mutex */
+	struct k_mutex zms_lock;
+	/** Flash device runtime structure */
+	const struct device *flash_device;
+	/** Flash memory parameters structure */
+	const struct flash_parameters *flash_parameters;
```

Comment:
I think you need to add a `#include <zephyr/drivers/flash.h>` on top of this header file (instead of adding it to the top of `zms.c`) otherwise `struct flash_parameters` might be undefined when this header file is parsed. Parsing would be successful only if the user adds `#include <zephyr/drivers/flash.h>` before `#include <zephyr/fs/zms.h>` in the code. But this is a public header, so it would be nice to be self contained IMO.